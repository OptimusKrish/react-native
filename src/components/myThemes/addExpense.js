import React, { Component } from "react";
import { StyleSheet } from "react-native";
import { Container, Content, Picker, Form, DatePicker, Item, Text, Input, Button, Spinner, Toast, View } from "native-base";
import Icon from 'react-native-vector-icons/FontAwesome';
import { DBUtils } from '../../constants/db';

export class AddExpense extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selected: "snacks",
      chosenDate: new Date(),
      expense: '',
      spinner: true,
      inputError: false,
      dateError: false
    };
    this.setDate = this.setDate.bind(this);
  }

  setDate(newDate) {
    this.setState({ chosenDate: newDate });
  }

  onValueChange(value) {
    this.setState({
      selected: value
    });
  }

  async saveExpense() {
    this.setState({ spinner: true });
    const data = {
      d: this.state.chosenDate.toString().substr(4, 12).trim(),
      cat: this.state.selected,
      amt: this.state.expense.trim(),
      expTimestamp: new Date(this.state.chosenDate).getTime(),
      timestamp: +new Date,
    }
    if(!data.amt) { 
      this.setState({ inputError: true, spinner: false, expense: '', dateError: false });
      return;
    }
    if (data.expTimestamp >= data.timestamp) {
      this.setState({ spinner: false, expense: '', dateError: true, inputError: false });
      Toast.show({
        type: "danger",
        text: "Cannot save future Expenses :-(",
        buttonText: "Okay",
        buttonTextStyle: { color: "#000000" },
        buttonStyle: { backgroundColor: "#f8f8ff" },
        duration: 5000
      });
      return;
    }
    await DBUtils.insert(data);
    this.setState({ spinner: false, expense: '', inputError: false, dateError: false });
    Toast.show({
      text: "Expense added successfully",
      buttonText: "Okay",
      buttonStyle: { backgroundColor: "#5cb85c" }
    });
  }

  async getExpense() {
    this.setState({ spinner: true });
    const exp = await DBUtils.select();
    this.setState({ spinner: false });
  }

  async init() {
    try {
      await DBUtils.create();
      this.setState({ spinner: false });
    } catch (err){
        console.log(err);
    }
  }

  render() {
    if (this.state.spinner) {
      return (
        <Container>
        <Spinner color='blue' />
        </Container>
      );
    }
    return (
      <Container style={{ borderRadius: 10, margin: 30 }}>
        <Content>
          <Form style = {{ marginLeft: 20, marginBottom: 30 }}>
          <Item rounded style= {{borderColor: "#696969"}}>
            <Picker
              mode="dropdown"
              selectedValue={this.state.selected}
              onValueChange={this.onValueChange.bind(this)}
            >
              <Picker.Item label="Snacks" value="snacks" />
              <Picker.Item label="Tea" value="tea" />
              <Picker.Item label="Cigrette" value="cigrette" />
              <Picker.Item label="Petrol" value="petrol" />
              <Picker.Item label="Food" value="food" />
            </Picker>
            </Item>
          </Form>
          <Item rounded style={[ this.state.dateError ? styles.dateError : styles.dateNormal ]}> 
            <Icon 
              name="calendar" 
              size={25} 
              style = {{marginLeft: 20, color: "#900"}}
              color="#900" 
            />
            <DatePicker
              minimumDate={new Date(2018, 1, 1)}
              maximumDate={new Date(2020, 12, 31)}
              locale={"en"}
              modalTransparent={false}
              animationType={"fade"}
              androidMode={"default"}
              placeHolderText={<Text style={{color: "#900"}}> {this.state.chosenDate.toString().substr(4, 12)} </Text>}
              textStyle={{ color: "green" }}
              placeHolderTextStyle={{ color: "#d3d3d3" }}
              onDateChange={this.setDate}
              disabled={false}
            />
          </Item>
            <Item rounded style = { [! this.state.inputError ? styles.normal : styles.error] }>
                <Input keyboardType = 'numeric' placeholder='Enter amount'
                 onChangeText={(expense) => this.setState({expense})} value={this.state.expense}
                 />
                 <View>
                 { 
                   !this.state.inputError && 
                   <Icon 
                   style = {{ fontSize: 15, marginRight: 20  }} 
                   name='rupee' 
                   />
                 }
                 { 
                   this.state.inputError && 
                   <Icon 
                   style = {{ fontSize: 18, marginRight: 20 }} 
                   name='warning' 
                   /> 
                 }
                 </View>
            </Item>
            <View style = {{ marginTop: 20, marginLeft: 90 }}>
                <Button style = {{margin: 10}} rounded success onPress={this.saveExpense.bind(this)}>
                    <Text>Save Expense</Text>
                </Button>
            </View>
        </Content>
      </Container>
    );
  }

  componentDidMount() {
    this.init();
  }
}

const styles = StyleSheet.create({
  error: {
    marginLeft: 20, 
    marginBottom: 30,
    borderColor: "#ff0000",
  },
  normal: {
    marginLeft: 20, 
    marginBottom: 30,
    borderColor: "#696969"
  },
  dateNormal: {
    marginLeft: 20, 
    marginBottom: 30, 
    borderColor: '#f5fffa'
  },
  dateError: {
    marginLeft: 20, 
    marginBottom: 30, 
    borderColor: "#ff0000",
  }
});