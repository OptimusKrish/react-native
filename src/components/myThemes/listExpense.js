import React, { Component } from "react";
import { Body, Text, Card, CardItem  } from 'native-base';
import Icon from 'react-native-vector-icons/FontAwesome';

export class ListExpense extends Component {
    
    render() {
        return(
                this.props.listExp.map(
                  (exp, i) => 
                  (
                    <Card style={{borderColor: "Red"}} key= {i}>
                    <CardItem header>
                      <Text>For {exp.Category}</Text>
                      { exp.sameDate && 
                        <Text style={{ position: 'absolute', right: 5 }}> <Icon name="calendar-times-o" size={17} color="#900" /> {exp.sameDate} {exp.Time}</Text>
                      }
                    </CardItem>
                    <CardItem>
                      <Body>
                        <Text>
                          Spent Rs.{exp.Amount}/-
                        </Text>
                      </Body>
                    </CardItem>
                    <CardItem footer>
                    { exp.sameDate && 
                        <Text>On {exp.Date}</Text>
                    }
                    { !exp.sameDate && 
                        <Text>On {exp.Date} <Icon name="clock-o" size={14} color="#900" /> {exp.Time}</Text>
                    }
                    </CardItem>
                  </Card>
                  )
                )
        )
    }
}

