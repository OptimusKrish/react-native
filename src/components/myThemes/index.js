
import { AddExpense } from './addExpense';
import { ListExpense } from './listExpense';

export {
    AddExpense,
    ListExpense
}