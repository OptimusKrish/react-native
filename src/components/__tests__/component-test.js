import 'react-native';
import React from 'react';
import { MonoText } from '../StyledText';
import { AddExpense } from '../myThemes/addExpense';
import { ListExpense } from '../myThemes/listExpense';
import renderer from 'react-test-renderer';

describe('Component test', () => {

  it('StyleText renders correctly', () => {
    const tree = renderer.create(<MonoText>Snapshot test!</MonoText>).toJSON();
  
    expect(tree).toMatchSnapshot();
  });

  it('AddExpense renders correctly', () => {
    const tree = renderer.create(<AddExpense/>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('ListExpense renders correctly', () => {
    const mockData =  [
      {
       "Amount": 15,
       "Category": "cigrette",
       "Date": "Feb 08 2019",
       "ID": 6,
       "Time": "21:39:34 PM",
       "Timestamp": 1549642174007,
       "expTimestamp": 1549642140003,
       "sameDate": false,
     },
     {
       "Amount": 23,
       "Category": "snacks",
       "Date": "Feb 08 2019",
       "ID": 1,
       "Time": "20:33:43 PM",
       "Timestamp": 1549638223978,
       "expTimestamp": 1549638213153,
       "sameDate": false,
     }
   ];
    const tree = renderer.create(<ListExpense listExp = {mockData}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  
});

