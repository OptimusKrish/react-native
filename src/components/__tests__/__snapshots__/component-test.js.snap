// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component test AddExpense function calls correctly 1`] = `undefined`;

exports[`Component test AddExpense function calls correctly 2`] = `undefined`;

exports[`Component test AddExpense function calls correctly 3`] = `
Promise {
  "_40": 0,
  "_55": null,
  "_65": 0,
  "_72": null,
}
`;

exports[`Component test AddExpense renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "height": 1334,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="blue"
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "height": 80,
      }
    }
  />
</View>
`;

exports[`Component test ListExpense renders correctly 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderBottomWidth": 0.5,
          "borderColor": "#ccc",
          "borderLeftWidth": 0.5,
          "borderRadius": 2,
          "borderRightWidth": 0.5,
          "borderTopWidth": 0.5,
          "elevation": 3,
          "flexWrap": "nowrap",
          "marginBottom": 5,
          "marginLeft": 2,
          "marginRight": 2,
          "marginTop": 5,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 1.5,
        },
        Object {
          "borderColor": "Red",
        },
      ]
    }
  >
    <View
      header={true}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderBottomWidth": null,
          "borderRadius": 2,
          "flexDirection": "row",
          "paddingBottom": 15,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 15,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
            "fontWeight": "600",
          }
        }
        uppercase={false}
      >
        For 
        cigrette
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 2,
          "flexDirection": "row",
          "paddingBottom": 10,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 10,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "alignSelf": "stretch",
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          uppercase={false}
        >
          Spent Rs.
          15
          /-
        </Text>
      </View>
    </View>
    <View
      footer={true}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderBottomWidth": null,
          "borderRadius": 2,
          "flexDirection": "row",
          "paddingBottom": 10,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
            "fontWeight": "600",
          }
        }
        uppercase={false}
      >
        On 
        Feb 08 2019
         
        <Text />
         
        21:39:34 PM
      </Text>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "borderBottomWidth": 0.5,
          "borderColor": "#ccc",
          "borderLeftWidth": 0.5,
          "borderRadius": 2,
          "borderRightWidth": 0.5,
          "borderTopWidth": 0.5,
          "elevation": 3,
          "flexWrap": "nowrap",
          "marginBottom": 5,
          "marginLeft": 2,
          "marginRight": 2,
          "marginTop": 5,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 1.5,
        },
        Object {
          "borderColor": "Red",
        },
      ]
    }
  >
    <View
      header={true}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderBottomWidth": null,
          "borderRadius": 2,
          "flexDirection": "row",
          "paddingBottom": 15,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 15,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
            "fontWeight": "600",
          }
        }
        uppercase={false}
      >
        For 
        snacks
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 2,
          "flexDirection": "row",
          "paddingBottom": 10,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 10,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "alignSelf": "stretch",
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 16,
            }
          }
          uppercase={false}
        >
          Spent Rs.
          23
          /-
        </Text>
      </View>
    </View>
    <View
      footer={true}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderBottomWidth": null,
          "borderRadius": 2,
          "flexDirection": "row",
          "paddingBottom": 10,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
            "fontWeight": "600",
          }
        }
        uppercase={false}
      >
        On 
        Feb 08 2019
         
        <Text />
         
        20:33:43 PM
      </Text>
    </View>
  </View>,
]
`;

exports[`Component test StyleText renders correctly 1`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "fontFamily": "space-mono",
      },
    ]
  }
>
  Snapshot test!
</Text>
`;
